lnrpc/signrpc/signer_server.go

```diff
@@ -0,0 +1,332 @@
+)
+
+// Server is a sub-server of the main RPC server: the signer RPC. This sub RPC
+// server allows external callers to access the full signing capabilities of
+// lnd. This allows callers to create custom protocols, external to lnd, even
+// backed by multiple distinct lnd across independent failure domains.
+type Server struct {
+	started  int32 // To be used atomically.
+	shutdown int32 // To be used atomically.
+
+	cfg *Config
+
+	quit chan struct{}
+}
+
+// A compile time check to ensure that Server fully implements the SignerServer
+// gRPC service.
+var _ SignerServer = (*Server)(nil)
+
+// fileExists reports whether the named file or directory exists.
+func fileExists(name string) bool {
+	if _, err := os.Stat(name); err != nil {
+		if os.IsNotExist(err) {
+			return false
+		}
+	}
+	return true
+}
+
+// New returns a new instance of the signrpc Signer sub-server. We also return
+// the set of permissions for the macaroons that we may create within this
+// method. If the macaroons we need aren't found in the filepath, then we'll
+// create them on start up. If we're unable to locate, or create the macaroons
+// we need, then we'll return with an error.
+func New(cfg *Config) (*Server, lnrpc.MacaroonPerms, error) {
+	// If the path of the signer macaroon wasn't generated, then we'll
+	// assume that it's found at the default network directory.
+	if cfg.SignerMacPath == "" {
+		cfg.SignerMacPath = filepath.Join(
+			cfg.NetworkDir, DefaultSignerMacFilename,
+		)
+	}
+
+	// Now that we know the full path of the signer macaroon, we can check
+	// to see if we need to create it or not.
+	macFilePath := cfg.SignerMacPath
+	if cfg.MacService != nil && !fileExists(macFilePath) {
+		log.Infof("Making macaroons for Signer RPC Server at: %v",
+			macFilePath)
+
+		// At this point, we know that the signer macaroon doesn't yet,
+		// exist, so we need to create it with the help of the main
+		// macaroon service.
+		signerMac, err := cfg.MacService.Oven.NewMacaroon(
+			context.Background(), bakery.LatestVersion, nil,
+			macaroonOps...,
+		)
+		if err != nil {
+			return nil, nil, err
+		}
+		signerMacBytes, err := signerMac.M().MarshalBinary()
+		if err != nil {
+			return nil, nil, err
+		}
+		err = ioutil.WriteFile(macFilePath, signerMacBytes, 0644)
+		if err != nil {
+			os.Remove(macFilePath)
+			return nil, nil, err
+		}
+	}
+
+	signerServer := &Server{
+		cfg:  cfg,
+		quit: make(chan struct{}),
+	}
+
+	return signerServer, macPermissions, nil
+}
+
+// Start launches any helper goroutines required for the rpcServer to function.
+//
+// NOTE: This is part of the lnrpc.SubServer interface.
+func (s *Server) Start() error {
+	if atomic.AddInt32(&s.started, 1) != 1 {
+		return nil
+	}
+
+	return nil
+}
+
+// Stop signals any active goroutines for a graceful closure.
+//
+// NOTE: This is part of the lnrpc.SubServer interface.
+func (s *Server) Stop() error {
+	if atomic.AddInt32(&s.shutdown, 1) != 1 {
+		return nil
+	}
+
+	close(s.quit)
```

Comment:
we don't use `quit` anywhere besides here atm. should be fine if both `Start` and `Stop` return nil? also means we can remove `started` and `stopped`